File: python/samples/02-agents/skills/subprocess_script_runner.py
# Copyright (c) Microsoft. All rights reserved.

"""Sample subprocess-based skill script runner.
Executes file-based skill scripts as local Python subprocesses.
This is provided for demonstration purposes only.
"""

from __future__ import annotations

import subprocess
import sys
from pathlib import Path
from typing import Any

from agent_framework import Skill, SkillScript


def subprocess_script_runner(skill: Skill, script: SkillScript, args: dict[str, Any] | None = None) -> str:
    """Run a skill script as a local Python subprocess.
    Resolves the script's absolute path from the skill directory, converts
    the ``args`` dict to CLI flags, and returns captured output.
    Args:
        skill: The skill that owns the script.
        script: The script to run.
        args: Optional arguments forwarded as CLI flags.
    Returns:
        The combined stdout/stderr output, or an error message.
    """
    if not skill.path:
        return f"Error: Skill '{skill.name}' has no directory path."
    if not script.path:
        return f"Error: Script '{script.name}' has no file path. Only file-based scripts can be executed locally."
    script_path = Path(skill.path) / script.path
    if not script_path.is_file():
        return f"Error: Script file not found: {script_path}"
    cmd = [sys.executable, str(script_path)]
    # Convert args dict to CLI flags
    if args:
        for key, value in args.items():
            if isinstance(value, bool):
                if value:
                    cmd.append(f"--{key}")
            elif value is not None:
                cmd.append(f"--{key}")
                cmd.append(str(value))
    try:
        result = subprocess.run(
            cmd,
            capture_output=True,
            text=True,
            timeout=30,
            cwd=str(script_path.parent),
        )
        output = result.stdout
        if result.stderr:
            output += f"\nStderr:\n{result.stderr}"
        if result.returncode != 0:
            output += f"\nScript exited with code {result.returncode}"
        return output.strip() or "(no output)"
    except subprocess.TimeoutExpired:
        return f"Error: Script '{script.name}' timed out after 30 seconds."
    except OSError as e:
        return f"Error: Failed to execute script '{script.name}': {e}"

File: dotnet/samples/02-agents/Harness/BuildYourOwnClaw/Claw_Step04_ProductionReady/ClawAgent/SubprocessScriptRunner.cs
// Copyright (c) Microsoft. All rights reserved.

using System.Diagnostics;
using System.Text.Json;
using Microsoft.Agents.AI;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;

namespace ClawAgent;

/// <summary>
/// Executes file-based skill scripts as local subprocesses.
/// </summary>
internal sealed class SubprocessScriptRunner
{
    private static readonly TimeSpan s_scriptTimeout = TimeSpan.FromSeconds(30);
    private readonly ILogger _logger;

    /// <summary>
    /// Initializes a new instance of the <see cref="SubprocessScriptRunner" /> class.
    /// </summary>
    public SubprocessScriptRunner(ILoggerFactory? loggerFactory = null)
    {
        this._logger = (loggerFactory ?? NullLoggerFactory.Instance).CreateLogger<SubprocessScriptRunner>();
    }

    /// <summary>
    /// Runs a skill script as a local subprocess.
    /// </summary>
    public async Task<object?> RunAsync(
        AgentFileSkill skill,
        AgentFileSkillScript script,
        JsonElement? arguments,
        IServiceProvider? serviceProvider,
        CancellationToken cancellationToken)
    {
        this._logger.LogDebug("Running script '{ScriptName}' from skill '{SkillName}'.", script.Name, skill.Frontmatter.Name);

        if (!File.Exists(script.FullPath))
        {
            this._logger.LogError("Script file not found for skill '{SkillName}': {ScriptPath}", skill.Frontmatter.Name, script.FullPath);
            return $"Error: Script file not found: {script.FullPath}";
        }

        string extension = Path.GetExtension(script.FullPath);
        string? interpreter = extension switch
        {
            ".py" => OperatingSystem.IsWindows() ? "python" : "python3",
            ".js" => "node",
            ".sh" => "bash",
            ".ps1" => "pwsh",
            _ => null,
        };

        var startInfo = new ProcessStartInfo
        {
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            UseShellExecute = false,
            CreateNoWindow = true,
            WorkingDirectory = Path.GetDirectoryName(script.FullPath) ?? ".",
        };

        if (interpreter is not null)
        {
            startInfo.FileName = interpreter;
            startInfo.ArgumentList.Add(script.FullPath);
        }
        else
        {
            startInfo.FileName = script.FullPath;
        }

        if (arguments is { ValueKind: JsonValueKind.Array } json)
        {
            foreach (var element in json.EnumerateArray())
            {
                if (element.ValueKind != JsonValueKind.String)
                {
                    throw new InvalidOperationException(
                        $"File-based skill scripts only accept string CLI arguments but received a JSON element of kind '{element.ValueKind}'. " +
                        "All array elements must be JSON strings.");
                }

                startInfo.ArgumentList.Add(element.GetString()!);
            }
        }
        else if (arguments?.ValueKind is not null and not JsonValueKind.Null and not JsonValueKind.Undefined)
        {
            throw new InvalidOperationException(
                $"Expected a JSON array of CLI arguments but received {arguments.Value.ValueKind}. " +
                "File-based skill scripts expect positional arguments as a JSON array of strings.");
        }

        using var timeoutCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
        timeoutCts.CancelAfter(s_scriptTimeout);
        CancellationToken runToken = timeoutCts.Token;

        Process? process = null;
        try
        {
            process = Process.Start(startInfo);
            if (process is null)
            {
                this._logger.LogError("Failed to start process for script '{ScriptName}' from skill '{SkillName}'.", script.Name, skill.Frontmatter.Name);
                return $"Error: Failed to start process for script '{script.Name}'.";
            }

            Task<string> outputTask = process.StandardOutput.ReadToEndAsync(runToken);
            Task<string> errorTask = process.StandardError.ReadToEndAsync(runToken);

            await process.WaitForExitAsync(runToken).ConfigureAwait(false);

            string output = await outputTask.ConfigureAwait(false);
            string error = await errorTask.ConfigureAwait(false);

            if (!string.IsNullOrEmpty(error))
            {
                if (process.ExitCode == 0)
                {
                    this._logger.LogWarning(
                        "Script '{ScriptName}' from skill '{SkillName}' succeeded but wrote to stderr:\n{Stderr}",
                        script.Name, skill.Frontmatter.Name, error.Trim());
                }

                output += $"\nStderr:\n{error}";
            }

            if (process.ExitCode != 0)
            {
                this._logger.LogError(
                    "Script '{ScriptName}' from skill '{SkillName}' exited with code {ExitCode}.{Stderr}",
                    script.Name,
                    skill.Frontmatter.Name,
                    process.ExitCode,
                    string.IsNullOrEmpty(error) ? string.Empty : $"\nStderr:\n{error.Trim()}");

                output += $"\nScript exited with code {process.ExitCode}";
            }

            string result = string.IsNullOrEmpty(output) ? "(no output)" : output.Trim();

            if (process.ExitCode == 0)
            {
                this._logger.LogInformation(
                    "Script '{ScriptName}' from skill '{SkillName}' completed successfully. Output:\n{Output}",
                    script.Name,
                    skill.Frontmatter.Name,
                    result);
            }

            return result;
        }
        catch (OperationCanceledException) when (!cancellationToken.IsCancellationRequested)
        {
            process?.Kill(entireProcessTree: true);
            this._logger.LogError(
                "Script '{ScriptName}' from skill '{SkillName}' timed out after {Timeout} seconds.",
                script.Name,
                skill.Frontmatter.Name,
                s_scriptTimeout.TotalSeconds);
            return $"Error: Script '{script.Name}' timed out after {s_scriptTimeout.TotalSeconds:0} seconds.";
        }
        catch (OperationCanceledException)
        {
            process?.Kill(entireProcessTree: true);
            throw;
        }
        catch (Exception ex)
        {
            this._logger.LogError(ex, "Failed to execute script '{ScriptName}' from skill '{SkillName}'.", script.Name, skill.Frontmatter.Name);
            return $"Error: Failed to execute script '{script.Name}': {ex.Message}";
        }
        finally
        {
            process?.Dispose();
        }
    }
}
